File: aicode/minicli3/src/minicli3/provider/glm.py
"""GLM-5 provider for MiniCLI3."""

from __future__ import annotations

import json
import time
from collections.abc import Generator
from typing import Any

import httpx


class ProviderError(Exception):
    pass


class GLMProvider:
    def __init__(
        self,
        api_key: str,
        base_url: str,
        model: str,
        timeout_ms: int,
        max_retries: int,
    ):
        if model != "glm-5":
            raise ProviderError("MiniCLI3 supports glm-5 only.")
        if not api_key:
            raise ProviderError("Missing API key. Set GLM_API_KEY or minicli3 config api_key.")

        self.api_key = api_key
        self.base_url = base_url.rstrip("/")
        self.model = model
        self.timeout_ms = timeout_ms
        self.max_retries = max_retries
        self.client = httpx.Client(
            timeout=timeout_ms / 1000,
            headers={
                "Authorization": f"Bearer {api_key}",
                "Content-Type": "application/json",
            },
        )

    def _url(self) -> str:
        return f"{self.base_url}/chat/completions"

    def chat(
        self,
        messages: list[dict[str, Any]],
        tools: list[dict[str, Any]] | None = None,
        stream: bool = True,
        temperature: float = 0.7,
        max_tokens: int = 4096,
    ) -> dict[str, Any] | Generator[dict[str, Any], None, None]:
        payload: dict[str, Any] = {
            "model": "glm-5",
            "messages": messages,
            "stream": stream,
            "temperature": temperature,
            "max_tokens": max_tokens,
        }
        if tools:
            payload["tools"] = tools

        last_error: Exception | None = None
        for attempt in range(self.max_retries + 1):
            try:
                if stream:
                    return self._stream_request(payload)
                response = self.client.post(self._url(), json=payload)
                return self._parse_response(response)
            except ProviderError as exc:
                last_error = exc
                if attempt == self.max_retries:
                    break
                time.sleep(2**attempt)
            except httpx.HTTPError as exc:
                last_error = ProviderError(f"http error: {exc}")
                if attempt == self.max_retries:
                    break
                time.sleep(2**attempt)

        raise last_error or ProviderError("unknown provider error")

    def _parse_response(self, response: httpx.Response) -> dict[str, Any]:
        if response.status_code == 429:
            raise ProviderError("rate limited (429)")
        if response.status_code >= 500:
            raise ProviderError(f"provider error ({response.status_code})")
        if response.status_code >= 400:
            raise ProviderError(f"request failed ({response.status_code}): {response.text}")

        try:
            payload = response.json()
        except json.JSONDecodeError as exc:
            raise ProviderError(f"invalid json response: {exc}") from exc

        if "error" in payload:
            err = payload["error"]
            msg = err.get("message", "provider returned error")
            raise ProviderError(msg)
        return payload

    def _stream_request(self, payload: dict[str, Any]) -> Generator[dict[str, Any], None, None]:
        with self.client.stream("POST", self._url(), json=payload) as response:
            if response.status_code != 200:
                self._parse_response(response)
                return
            for line in response.iter_lines():
                if not line:
                    continue
                text = line.strip()
                if not text.startswith("data: "):
                    continue
                data = text[6:]
                if data == "[DONE]":
                    break
                yield json.loads(data)

    def close(self) -> None:
        self.client.close()

File: aicode/minicli5/src/provider/glm.ts
import { BaseProvider } from "./base.js";
import { withRetry } from "./retry.js";
import { log } from "../utils/logger.js";
import type { ChatRequest, ChatResponse, StreamChunk, AppConfig } from "../types.js";

export class GlmProvider extends BaseProvider {
  private apiBase: string;
  private apiKey: string;
  private model: string;
  private maxRetries: number;
  private defaultMaxTokens: number;
  private defaultTemperature: number;

  constructor(config: AppConfig) {
    super();
    this.apiBase = config.api_base;
    this.apiKey = config.api_key;
    this.model = config.model;
    this.maxRetries = config.max_retries;
    this.defaultMaxTokens = config.max_tokens;
    this.defaultTemperature = config.temperature;
  }

  private buildBody(req: ChatRequest): Record<string, unknown> {
    const body: Record<string, unknown> = {
      model: req.model ?? this.model,
      messages: req.messages,
      max_tokens: req.max_tokens ?? this.defaultMaxTokens,
      temperature: req.temperature ?? this.defaultTemperature,
      stream: req.stream ?? false,
    };
    if (req.tools && req.tools.length > 0) {
      body.tools = req.tools;
      body.tool_choice = "auto";
    }
    return body;
  }

  private get headers(): Record<string, string> {
    return {
      "Content-Type": "application/json",
      Authorization: `Bearer ${this.apiKey}`,
    };
  }

  async chat(req: ChatRequest): Promise<ChatResponse> {
    return withRetry(async () => {
      const body = this.buildBody({ ...req, stream: false });
      const res = await fetch(`${this.apiBase}/chat/completions`, {
        method: "POST",
        headers: this.headers,
        body: JSON.stringify(body),
      });

      if (!res.ok) {
        const text = await res.text();
        log("error", "GLM API error", { status: res.status, body: text });
        const err = new Error(`GLM API error ${res.status}: ${text}`);
        (err as any).status = res.status;
        throw err;
      }

      return (await res.json()) as ChatResponse;
    }, { maxRetries: this.maxRetries });
  }

  async *chatStream(req: ChatRequest): AsyncIterable<StreamChunk> {
    const body = this.buildBody({ ...req, stream: true });

    const res = await withRetry(async () => {
      const r = await fetch(`${this.apiBase}/chat/completions`, {
        method: "POST",
        headers: this.headers,
        body: JSON.stringify(body),
      });
      if (!r.ok) {
        const text = await r.text();
        log("error", "GLM stream API error", { status: r.status, body: text });
        const err = new Error(`GLM API error ${r.status}: ${text}`);
        (err as any).status = r.status;
        throw err;
      }
      return r;
    }, { maxRetries: this.maxRetries });

    const reader = res.body?.getReader();
    if (!reader) throw new Error("No response body");

    const decoder = new TextDecoder();
    let buffer = "";

    try {
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        buffer += decoder.decode(value, { stream: true });
        const lines = buffer.split("\n");
        buffer = lines.pop() ?? "";

        for (const line of lines) {
          const trimmed = line.trim();
          if (!trimmed || !trimmed.startsWith("data:")) continue;

          const data = trimmed.slice(5).trim();
          if (data === "[DONE]") return;

          try {
            yield JSON.parse(data) as StreamChunk;
          } catch {
            log("warn", "Failed to parse SSE chunk", { data });
          }
        }
      }
    } finally {
      reader.releaseLock();
    }
  }
}
